<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Stray::Pibble • stray</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Stray::Pibble">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stray</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="https://github.com/jsilve24/stray/wiki/Installation-Details">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction-to-stray.html">Intro to stray through stray::pibble</a>
    </li>
    <li>
      <a href="../articles/non-linear-models.html">Non-Linear Modeling with stray::basset</a>
    </li>
    <li>
      <a href="../articles/picking_priors.html">Tips on Specifying Priors</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://jsilve24.github.io/driver/">driver</a>
    </li>
    <li>
      <a href="https://bioconductor.org/packages/release/bioc/html/philr.html">philr</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/inschool4life">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/jsilve24/stray">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Introduction to Stray::Pibble</h1>
                        <h4 class="author">Justin Silverman</h4>
            
            <h4 class="date">2019-06-22</h4>
      
      
      <div class="hidden name"><code>introduction-to-stray.Rmd</code></div>

    </div>

    
    
<div id="an-introduction-to-stray" class="section level1">
<h1 class="hasAnchor">
<a href="#an-introduction-to-stray" class="anchor"></a>An introduction to <em>stray</em>
</h1>
<p><em>stray</em> <span class="citation">(Silverman et al. 2019)</span> is a loose acronym for <strong>“(Bayesian) Multinomial Logistic-Normal Models”</strong>. In particular the development of <em>stray</em> stems from the need for fast inference for time-invariant MALLARD models<span class="citation">(Silverman et al. 2018)</span>. <em>stray</em> is very fast! It uses closed form solutions for model gradients and Hessians written in C++ to preform <a href="https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation">MAP estimation</a> in combination with parameter uncertainty estimation using a <a href="http://www.sumsar.net/blog/2013/11/easy-laplace-approximation/">Laplace Approximation</a>. One of the main models in <em>stray</em> is the function <em>pibble</em> which fits a Multinomial Logistic-Normal <strong>Linear Regerssion</strong> model.</p>
<p><strong>So what is a <em>stray</em> model exactly?</strong> First let me give the broad description from 10,000ft up: Basically its a model for multinomial count data (e.g., each sample contains the counts of <span class="math inline">\(D\)</span> “types of things”). Importantly, unlike the more common Poisson count models, the multinomial models a “competition to be counted” (i.e., cases in which counting more of one type of thing means that I have less resources available to count other types of things).</p>
<p>This may seem vague so let me give an example. Pretend there is a ball pit with red, green, and blue balls. Pretend that the ball pit is very large and I don’t know the total number of balls in the ball pit, yet I want to say something about the relative number of red, blue, and green balls in the pit. One way I may choose to measure the ball pit is by grabbing an armful of balls and counting the number of balls of each color (e.g., in one armful I may collect 5 red, 3 blue, and 6 green). My arms can only contain so many balls (in this example about 14) and so if I were to have (randomly) gotten another green ball in my armful (making 7 total) I would likely not have been able to measure one of the red or blue balls; hence the “competition to be counted”. It turns out that this type of sampling occurs all the time in many situations (Wikipedia has an example with <a href="https://en.wikipedia.org/wiki/Multinomial_distribution#Example">political polling</a>). Perhaps one of the most notable examples of this type of count data occurs with modern high-throughput sequencing studies such as 16S rRNA studies to profile microbial communities or bulk/single-cell RNA-seq studies to study expression profiles of cells. In all cases, transcripts are sequenced and the number of different types of transcripts are counted. The important part is that sequencing only samples a small portion of the total genetic material available and leads to similar competition to be counted.</p>
<div id="the-pibble-model" class="section level2">
<h2 class="hasAnchor">
<a href="#the-pibble-model" class="anchor"></a>The <em>pibble</em> model</h2>
<p><em>Pibble</em> is one type of <em>stray</em> model. In particular its a <em>stray</em> model for multivariate linear regression.</p>
<p>Let <span class="math inline">\(Y\)</span> denote an <span class="math inline">\(D\times N\)</span> matrix of counts. Let us denote the <span class="math inline">\(j\)</span>-th column of <span class="math inline">\(Y\)</span> as <span class="math inline">\(Y_j\)</span>. Thus each “sample” in the dataset is a measurement of the relative amount of <span class="math inline">\(D\)</span> “types of things”. Suppose we also have have covariate information in the form of a <span class="math inline">\(Q\times N\)</span> matrix <span class="math inline">\(X\)</span>.</p>
<p>The following is the pibble model including likelihood and priors: <span class="math display">\[
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j \right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\eta_j &amp;\sim N(\Lambda X_j, \Sigma) \\
\Lambda &amp;\sim  MN_{(D-1) \times Q}(\Theta, \Sigma, \Gamma) \\
\Sigma &amp;\sim W^{-1}(\Xi, \upsilon) 
\end{align}
\]</span> Here <span class="math inline">\(MN_{(D-1) \times Q}\)</span> denotes a <a href="https://en.wikipedia.org/wiki/Matrix_normal_distribution">Matrix Normal distribution</a> for a matrix <span class="math inline">\(\Lambda\)</span> of regression coefficients of dimension <span class="math inline">\((D-1)\times Q\)</span>. Essentially you can think of the Matrix normal as having two covariance matrices one describing the covariation between the rows of <span class="math inline">\(\Lambda\)</span> (<span class="math inline">\(\Sigma\)</span>) and another describing the covariation of the columns of <span class="math inline">\(\Lambda\)</span> (<span class="math inline">\(\Gamma\)</span>). and <span class="math inline">\(W^{-1}\)</span> refers to the <a href="https://en.wikipedia.org/wiki/Inverse-Wishart_distribution">Inverse Wishart distribution</a> (which is a common distribution over covariance matrices). The line <span class="math inline">\(\pi_j = \phi^{-1}(\eta_j)\)</span> represents a transformation between the parameters <span class="math inline">\(\pi_j\)</span> which exist on a simplex (e.g., <span class="math inline">\(\pi_j\)</span> must sum to 1) and the transformed parameters <span class="math inline">\(\eta_j\)</span> that exist in real space. In particular we define <span class="math inline">\(\phi^{-1}\)</span> to be the <a href="http://www.sediment.uni-goettingen.de/staff/tolosana/extra/CoDaNutshell.pdf">inverse additive log ratio transform</a> (which conversely implies that <span class="math inline">\(\eta_j = ALR(\pi_j)\)</span>) also known as the identified softmax transform (as it is more commonly known in the Machine Learning community). While I will say more on this later in this tutorial, one thing to know is that I have the model implemented using the ALR transform as it is computationally simple and fast; the results of the model can be viewed as if any number of transforms had been used (instead of the ALR) including the isometric log-ratio transform, or the centered log-ratio transform.</p>
<p>Before moving on, I would like to give <strong>a more intuitive description of <em>pibble</em></strong>. Essentially the main modeling component of <em>pibble</em> is the third equation above (<span class="math inline">\(\eta_j \sim N(\Lambda X_j, \Sigma)\)</span>) which is just a multivariate linear model. That is, <span class="math inline">\(X\)</span> are your covariates (which can be continuous, discrete, binary, etc…), and <span class="math inline">\(\Sigma\)</span> is the covariance matrix for the regression residuals.</p>
</div>
</div>
<div id="example-analysis-of-microbiome-data" class="section level1">
<h1 class="hasAnchor">
<a href="#example-analysis-of-microbiome-data" class="anchor"></a>Example analysis of microbiome data</h1>
<p>This analysis is the same as that presented in the <em>stray</em> manuscript <span class="citation">(Silverman et al. 2019)</span>. I will reanalyze a previously published study comparing microbial composition in the terminal ileum of subjects with Crohn’s Disease (CD) to healthy controls <span class="citation">(Gevers et al. 2014)</span>. To do this I will fit a pibble model using CD status, inflammation status and age as covariates (plus a constant intercept term).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MicrobeDS)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(phyloseq)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(stray)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">899</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"RISK_CCFA"</span>)
<span class="co"># drop low abundant taxa and samples</span>
dat &lt;-<span class="st"> </span>RISK_CCFA <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/subset_samples-methods">subset_samples</a></span>(disease_stat<span class="op">!=</span><span class="st">"missing"</span>, 
                 immunosup<span class="op">!=</span><span class="st">"missing"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/subset_samples-methods">subset_samples</a></span>(diagnosis <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"no"</span>, <span class="st">"CD"</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/subset_samples-methods">subset_samples</a></span>(steroids<span class="op">==</span><span class="st">"false"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/subset_samples-methods">subset_samples</a></span>(antibiotics<span class="op">==</span><span class="st">"false"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/subset_samples-methods">subset_samples</a></span>(biologics<span class="op">==</span><span class="st">"false"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/subset_samples-methods">subset_samples</a></span>(biopsy_location<span class="op">==</span><span class="st">"Terminal ileum"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/tax_glom">tax_glom</a></span>(<span class="st">"Family"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/prune_samples-methods">prune_samples</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/sample_sums">sample_sums</a></span>(.) <span class="op">&gt;=</span><span class="st"> </span><span class="dv">5000</span>,.) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/filter_taxa">filter_taxa</a></span>(<span class="cf">function</span>(x) <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(x <span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.10</span><span class="op">*</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(x), <span class="ot">TRUE</span>)</code></pre></div>
<p>Create Design Matrix and OTU Table</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_dat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/as.data.frame">as.data.frame</a></span>(<span class="kw">as</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/sample_data-methods">sample_data</a></span>(dat),<span class="st">"matrix"</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">age =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(age)),
         <span class="dt">diagnosis =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/relevel">relevel</a></span>(diagnosis, <span class="dt">ref=</span><span class="st">"no"</span>), 
         <span class="dt">disease_stat =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/relevel">relevel</a></span>(disease_stat, <span class="dt">ref=</span><span class="st">"non-inflamed"</span>))
X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/transpose-methods">t</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/model.matrix">model.matrix</a></span>(<span class="op">~</span>diagnosis <span class="op">+</span><span class="st"> </span>disease_stat<span class="op">+</span>age, <span class="dt">data=</span>sample_dat))
Y &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/otu_table-methods">otu_table</a></span>(dat)

<span class="co"># Investigate X and Y look like</span>
X[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;                             1        2     3        4     5</span>
<span class="co">#&gt; (Intercept)           1.00000  1.00000  1.00  1.00000  1.00</span>
<span class="co">#&gt; diagnosisCD           1.00000  1.00000  1.00  1.00000  1.00</span>
<span class="co">#&gt; disease_statinflamed  0.00000  1.00000  1.00  1.00000  1.00</span>
<span class="co">#&gt; age                  15.16667 14.33333 15.75 13.58333 15.75</span>
Y[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt; OTU Table:          [5 taxa and 5 samples]</span>
<span class="co">#&gt;                      taxa are rows</span>
<span class="co">#&gt;         1939.SKBTI.0175 1939.SKBTI047 1939.SKBTI051 1939.SKBTI063</span>
<span class="co">#&gt; 4442127               0             9             0            14</span>
<span class="co">#&gt; 74305                 1             2            35             1</span>
<span class="co">#&gt; 663573               36             1             0             2</span>
<span class="co">#&gt; 2685602              10           264           211           276</span>
<span class="co">#&gt; 4339819               0            37            42            70</span>
<span class="co">#&gt;         1939.SKBTI072</span>
<span class="co">#&gt; 4442127             2</span>
<span class="co">#&gt; 74305               0</span>
<span class="co">#&gt; 663573              1</span>
<span class="co">#&gt; 2685602            83</span>
<span class="co">#&gt; 4339819            22</span></code></pre></div>
<p>Next specify priors. We are going to start by specifying a prior on the covariance between log-ratios <span class="math inline">\(\Sigma\)</span>. I like to do this by thinking about a prior on the covariance between taxa on the log-scale (i.e., between the log of their absolute abundances not the log-ratios). I will refer to this covariance on log-absolute abundances <span class="math inline">\(\Omega\)</span>. For example, here I will build a prior that states that the mean of <span class="math inline">\(\Omega\)</span> is the identity matrix <span class="math inline">\(I_D\)</span>. From From <span class="citation">Aitchison (1986)</span>, we know that if we assume that the taxa have a covariance <span class="math inline">\(\Omega\)</span> in terms of log-absolute abundance then their correlation in the <span class="math inline">\(\text{ALR}_D\)</span> is given by <span class="math display">\[ \Sigma = G \Omega G^T \]</span> where <span class="math inline">\(G\)</span> is a <span class="math inline">\(D-1 \times D\)</span> matrix given by <span class="math inline">\(G = [I_{D-1}; -1_{D-1}]\)</span> (i.e., <span class="math inline">\(G\)</span> is the <span class="math inline">\(\text{ALR}_D\)</span> contrast matrix). Additionally, we know that the Inverse Wishart mode is given by <span class="math inline">\(\frac{\Xi}{\upsilon + D}\)</span>. Finally, note that <span class="math inline">\(\upsilon\)</span> essentially controls our uncertainty in <span class="math inline">\(\Sigma\)</span> about this prior mean. Here I will take <span class="math inline">\(\upsilon = D+3\)</span>. This then gives us <span class="math inline">\(\Xi = (\upsilon - D) GIG^T\)</span>. We scale <span class="math inline">\(\Xi\)</span> by a factor of 1/2 to make <span class="math inline">\(Tr(\Xi)=D-1\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">upsilon &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/ntaxa-methods">ntaxa</a></span>(dat)<span class="op">+</span><span class="dv">3</span> 
Omega &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/ntaxa-methods">ntaxa</a></span>(dat))
G &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/ntaxa-methods">ntaxa</a></span>(dat)<span class="op">-</span><span class="dv">1</span>), <span class="op">-</span><span class="dv">1</span>)
Xi &lt;-<span class="st"> </span>(upsilon<span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/ntaxa-methods">ntaxa</a></span>(dat))<span class="op">*</span>G<span class="op">%*%</span>Omega<span class="op">%*%</span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/transpose-methods">t</a></span>(G)</code></pre></div>
<p>Finally I specify my priors for <span class="math inline">\(\Theta\)</span> (mean of <span class="math inline">\(\Lambda\)</span>) and <span class="math inline">\(\Gamma\)</span> (covariance between columns of <span class="math inline">\(\Lambda\)</span>; i.e., covariance between the covariates). I will center my prior for <span class="math inline">\(\Lambda\)</span> about zero, and assume that the covariates are independent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Theta &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dv">0</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/ntaxa-methods">ntaxa</a></span>(dat)<span class="op">-</span><span class="dv">1</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(X))
Gamma &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(X))</code></pre></div>
<p>I strongly recommend users perform prior predictive checks to make sure their priors make sense to them. <em>stray</em> makes this easy, all the main fitting functions (e.g., <code>pibble</code>) will automatically sample from the prior predictive distribution if <code>Y</code> is left as <code>NULL</code> (e.g., without data your posterior is just your prior).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">priors &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pibble_fit.html">pibble</a></span>(<span class="ot">NULL</span>, X, upsilon, Theta, Gamma, Xi)  
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(priors)
<span class="co">#&gt;  pibblefit Object (Priors Only): </span>
<span class="co">#&gt;   Number of Samples:      250 </span>
<span class="co">#&gt;   Number of Categories:       49 </span>
<span class="co">#&gt;   Number of Covariates:       4 </span>
<span class="co">#&gt;   Number of Posterior Samples:    2000 </span>
<span class="co">#&gt;   Contains Samples of Parameters:Eta  Lambda  Sigma</span>
<span class="co">#&gt;   Coordinate System:      alr, reference category: 49</span></code></pre></div>
<p>The main fitting functions in the <em>stray</em> package output special fit objects (e.g., <code>pibble</code> outputs an object of class <code>pibblefit</code>). These fit objects are just lists with some extra metadata that allows special method dispatch. For example, if you call print on a <code>pibblefit</code> object you will get a nice summary of what is in the object.</p>
<p><em>Note:</em> Currently, the function <code>pibble</code> takes expects inputs and outputs in the “default” coordinate system; this is simply the ALR coordinate system where the last category (49 above) is taken as reference (this will be generalized in future versions). More specifically for a vector <span class="math inline">\(x\)</span> representing the proportions of categories <span class="math inline">\(\{1, \dots, D\}\)</span> we can write <span class="math display">\[x^* = \left( \log \frac{x_1}{x_D}, \dots, \log \frac{x_{D-1}}{x_D}\right).\]</span> As mentioned above however, I have designed <em>stray</em> to work with many different coordinate systems including the ALR (with respect to any category), CLR, ILR, or proportions. To help transform things between these coordinate systems I have written a series of transformation functions that transform any <code>pibblefit</code> object into a desired coordinate system. Importantly, <code>pibblefit</code> objects keep track of what coordinate system they are currently in so as a user you only need to specify the coordinate system that you want to change into. Keep in mind that covariance matrices cannot be represented in proportions and so visualizations or summaries based on covariance matrices will be suppressed when <code>pibblefit</code> objects are in the proportions coordinate system. As an example, lets look at viewing a summary of the prior for <span class="math inline">\(\Lambda\)</span> with respect to the CLR coordinate system<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">priors &lt;-<span class="st"> </span><span class="kw"><a href="../reference/stray_transforms.html">to_clr</a></span>(priors)  
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(priors, <span class="dt">pars=</span><span class="st">"Lambda"</span>)  
<span class="co">#&gt; $Lambda</span>
<span class="co">#&gt; # A tibble: 196 x 9</span>
<span class="co">#&gt;    Parameter coord covariate  p2.5    p25      p50        mean   p75 p97.5</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;       &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 Lambda        1         1 -1.74 -0.528  0.0421   0.0509     0.596  2.09</span>
<span class="co">#&gt;  2 Lambda        1         2 -2.08 -0.567  0.0143   0.00199    0.575  1.96</span>
<span class="co">#&gt;  3 Lambda        1         3 -2.06 -0.620 -0.0102  -0.0373     0.532  1.98</span>
<span class="co">#&gt;  4 Lambda        1         4 -1.90 -0.554 -0.00469 -0.00205    0.549  2.01</span>
<span class="co">#&gt;  5 Lambda        2         1 -1.86 -0.564 -0.0249   0.00991    0.535  2.13</span>
<span class="co">#&gt;  6 Lambda        2         2 -1.89 -0.572 -0.00610 -0.00000782 0.580  2.02</span>
<span class="co">#&gt;  7 Lambda        2         3 -2.14 -0.570 -0.0388  -0.0247     0.518  1.93</span>
<span class="co">#&gt;  8 Lambda        2         4 -2.00 -0.536  0.0240   0.0165     0.589  1.95</span>
<span class="co">#&gt;  9 Lambda        3         1 -2.02 -0.584  0.0224   0.0138     0.620  2.04</span>
<span class="co">#&gt; 10 Lambda        3         2 -2.05 -0.542 -0.00311 -0.00502    0.581  1.86</span>
<span class="co">#&gt; # … with 186 more rows</span></code></pre></div>
<p>By default the <code>summary</code> function returns a list (with possible elements <code>Lambda</code>, <code>Sigma</code>, and <code>Eta</code>) summarizing each posterior parameter based on quantiles and mean (e.g., p2.5 is the 0.025 percentile of the posterior distribution). As this type of table may be hard to take in due to how large it is, <code>pibblefit</code> objects also come with a default plotting option for each of the parameters. Also the returned plot objects are <code>ggplot</code> objects so normal <code>ggplot2</code> commands work on them. Before doing that though we are going to use one of the <code>names</code> functions for <code>pibblefit</code> objects to provide some more specific names for the covariates (helpful when we then plot).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/name_dims.html">names_covariates</a></span>(priors) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(X)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(priors, <span class="dt">par=</span><span class="st">"Lambda"</span>) <span class="op">+</span><span class="st"> </span>ggplot2<span class="op">::</span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="op">-</span><span class="dv">10</span>, <span class="dv">10</span>))  </code></pre></div>
<p><img src="introduction-to-stray_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<p>This looks fairly reasonable to me. So I am going to go ahead and fit the model with data. <code>stray</code> provides a helper method called <code>refit</code> that we will use to avoid passing prior parameters again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">priors<span class="op">$</span>Y &lt;-<span class="st"> </span>Y <span class="co"># remember pibblefit objects are just lists</span>
posterior &lt;-<span class="st"> </span><span class="kw"><a href="../reference/refit.html">refit</a></span>(priors, <span class="dt">optim_method=</span><span class="st">"adam"</span>)</code></pre></div>
<p>Unlike the main <em>pibble</em> function, the <code>refit</code> method can be called on objects in any coordinate system and all transformations to and from the default coordinate system are handled internally<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a>. This is one nice thing about using the <code>refit</code> method. That said, new objects added to the <code>pibblefit</code> object need to be added in the proper coordinates For example, if we wanted to replace our prior for <span class="math inline">\(\Xi\)</span> for an object in CLR coordinates, we would had to transform our prior for <code>Xi</code> to CLR coordinates before adding it to the <code>priors</code> object.</p>
<p>Now I are also going to add in the taxa names to make it easier to interpret the results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tax &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/tax_table-methods">tax_table</a></span>(dat)[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Class"</span>, <span class="st">"Family"</span>)]
tax &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(tax, <span class="dv">1</span>, paste, <span class="dt">collapse=</span><span class="st">"_"</span>)
<span class="kw"><a href="../reference/name_dims.html">names_categories</a></span>(posterior) &lt;-<span class="st"> </span>tax</code></pre></div>
<p>Before doing anything else lets look at the posterior predictive distribution to assess model fit. This can be accessed through the method <code>ppc</code><a href="#fn3" class="footnoteRef" id="fnref3"><sup>3</sup></a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ppc.html">ppc</a></span>(posterior) <span class="op">+</span><span class="st"> </span>ggplot2<span class="op">::</span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="dt">ylim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">30000</span>))</code></pre></div>
<p><img src="introduction-to-stray_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>There are a few things to note about this plot. First, when zoomed out like this it looks it is hard to make much of it. This is a fairly large dataset we are analyzing and its hard to view an uncertainty interval; in this case its plotting the median and 95% confidence interval in grey and black and the observed counts in green. <em>stray</em> also has a simpler function that summarizes the posterior predictive check.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ppc_summary.html">ppc_summary</a></span>(posterior)
<span class="co">#&gt; Proportions of Observations within 95% Credible Interval: 0.9902857</span></code></pre></div>
<p>Here we see that the model appears to be fitting well (at least based on the posterior predictive check) and that only about 1.5% of observations fall outside of the 95% posterior predictive density (this is good).</p>
<p>Some readers will look at the above <code>ppc</code> plots and think “looks like over-fitting”. However, note that there are two ways of using <code>ppc</code>. One is to predict the counts based on the samples of <span class="math inline">\(\eta\)</span> (Eta; as we did above); the other is to predict “from scratch” that is to predict starting form the posterior samples of <span class="math inline">\(\Lambda\)</span> (Lambda) then sampling <span class="math inline">\(\eta\)</span> and only then sampling <span class="math inline">\(Y\)</span>. This later functionality can be accessed by also passing the parameters <code>from_scratch=TRUE</code> to the <code>ppc</code> function. Note: these two posterior predictive checks have different meanings, one is not better than the other.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ppc.html">ppc</a></span>(posterior, <span class="dt">from_scratch=</span><span class="ot">TRUE</span>) <span class="op">+</span>ggplot2<span class="op">::</span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span>(<span class="dt">ylim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">30000</span>))</code></pre></div>
<p><img src="introduction-to-stray_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ppc_summary.html">ppc_summary</a></span>(posterior, <span class="dt">from_scratch=</span><span class="ot">TRUE</span>)
<span class="co">#&gt; Proportions of Observations within 95% Credible Interval: 0.972</span></code></pre></div>
<p>Now we are going to finally look at the posterior distribution of our regression parameters, but because there are so many we will focus on just those that have a 95% credible interval not including zero (i.e., those that the model is fairly certain are non-zero). We are also going to ignore the intercept term and just look at parameters associated with age and disease status.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">posterior_summary &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(posterior, <span class="dt">pars=</span><span class="st">"Lambda"</span>)<span class="op">$</span>Lambda
focus &lt;-<span class="st"> </span>posterior_summary[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sign">sign</a></span>(posterior_summary<span class="op">$</span>p2.<span class="dv">5</span>) <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sign">sign</a></span>(posterior_summary<span class="op">$</span>p97.<span class="dv">5</span>),]
focus &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(focus<span class="op">$</span>coord)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(posterior, <span class="dt">par=</span><span class="st">"Lambda"</span>, <span class="dt">focus.coord =</span> focus, <span class="dt">focus.cov =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(X)[<span class="dv">2</span><span class="op">:</span><span class="dv">4</span>])</code></pre></div>
<p><img src="introduction-to-stray_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>The first, and most obvious ting to notice is that the covariate <code>age</code> has pretty much no effect at all, whatever effect it may have is incredibly weak. So we are going to remove age from the plot and just look at those coordinates with non-zero effect for diagnosis CD</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">posterior_summary &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(posterior_summary, covariate<span class="op">==</span><span class="st">"diagnosisCD"</span>) 
focus &lt;-<span class="st"> </span>posterior_summary[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sign">sign</a></span>(posterior_summary<span class="op">$</span>p2.<span class="dv">5</span>) <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sign">sign</a></span>(posterior_summary<span class="op">$</span>p97.<span class="dv">5</span>),]
focus &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(focus<span class="op">$</span>coord)

<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/tax_table-methods">tax_table</a></span>(dat)[<span class="kw"><a href="https://www.rdocumentation.org/packages/phyloseq/topics/taxa_names-methods">taxa_names</a></span>(dat)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(<span class="kw"><a href="../reference/name_dims.html">names_coords</a></span>(posterior) <span class="op">%in%</span><span class="st"> </span>focus)]]
<span class="co">#&gt; Taxonomy Table:     [13 taxa by 7 taxonomic ranks]:</span>
<span class="co">#&gt;         Kingdom    Phylum           Class                  </span>
<span class="co">#&gt; 74305   "Bacteria" "Proteobacteria" "Epsilonproteobacteria"</span>
<span class="co">#&gt; 4449236 "Bacteria" "Proteobacteria" "Betaproteobacteria"   </span>
<span class="co">#&gt; 1105919 "Bacteria" "Proteobacteria" "Betaproteobacteria"   </span>
<span class="co">#&gt; 4477696 "Bacteria" "Proteobacteria" "Gammaproteobacteria"  </span>
<span class="co">#&gt; 4448331 "Bacteria" "Proteobacteria" "Gammaproteobacteria"  </span>
<span class="co">#&gt; 4154872 "Bacteria" "Bacteroidetes"  "Flavobacteriia"       </span>
<span class="co">#&gt; 4452538 "Bacteria" "Fusobacteria"   "Fusobacteriia"        </span>
<span class="co">#&gt; 341322  "Bacteria" "Firmicutes"     "Bacilli"              </span>
<span class="co">#&gt; 1015143 "Bacteria" "Firmicutes"     "Bacilli"              </span>
<span class="co">#&gt; 176318  "Bacteria" "Firmicutes"     "Clostridia"           </span>
<span class="co">#&gt; 1788466 "Bacteria" "Firmicutes"     "Clostridia"           </span>
<span class="co">#&gt; 1896700 "Bacteria" "Firmicutes"     "Clostridia"           </span>
<span class="co">#&gt; 191718  "Bacteria" "Firmicutes"     "Erysipelotrichi"      </span>
<span class="co">#&gt;         Order                Family                  Genus Species</span>
<span class="co">#&gt; 74305   "Campylobacterales"  "Helicobacteraceae"     NA    NA     </span>
<span class="co">#&gt; 4449236 "Burkholderiales"    "Alcaligenaceae"        NA    NA     </span>
<span class="co">#&gt; 1105919 "Burkholderiales"    "Oxalobacteraceae"      NA    NA     </span>
<span class="co">#&gt; 4477696 "Pasteurellales"     "Pasteurellaceae"       NA    NA     </span>
<span class="co">#&gt; 4448331 "Enterobacteriales"  "Enterobacteriaceae"    NA    NA     </span>
<span class="co">#&gt; 4154872 "Flavobacteriales"   "[Weeksellaceae]"       NA    NA     </span>
<span class="co">#&gt; 4452538 "Fusobacteriales"    "Fusobacteriaceae"      NA    NA     </span>
<span class="co">#&gt; 341322  "Turicibacterales"   "Turicibacteraceae"     NA    NA     </span>
<span class="co">#&gt; 1015143 "Gemellales"         "Gemellaceae"           NA    NA     </span>
<span class="co">#&gt; 176318  "Clostridiales"      "Christensenellaceae"   NA    NA     </span>
<span class="co">#&gt; 1788466 "Clostridiales"      "Lachnospiraceae"       NA    NA     </span>
<span class="co">#&gt; 1896700 "Clostridiales"      "Peptostreptococcaceae" NA    NA     </span>
<span class="co">#&gt; 191718  "Erysipelotrichales" "Erysipelotrichaceae"   NA    NA</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(posterior, <span class="dt">par=</span><span class="st">"Lambda"</span>, <span class="dt">focus.coord =</span> focus, <span class="dt">focus.cov =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(X)[<span class="dv">2</span>])</code></pre></div>
<p><img src="introduction-to-stray_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
</div>
<div id="more-technical-details" class="section level1">
<h1 class="hasAnchor">
<a href="#more-technical-details" class="anchor"></a>More Technical Details</h1>
<div id="a-few-notes-on-model-inference-and-parameter-collapsing" class="section level2">
<h2 class="hasAnchor">
<a href="#a-few-notes-on-model-inference-and-parameter-collapsing" class="anchor"></a>A few notes on model inference and parameter collapsing</h2>
<p>Along with some algorithmic speed-ups enabled by the C++ Eigen library <em>stray</em> uses conjugate priors for the regression component of the model allowing the last three lines of the model to be collapsed into 1 line. After this the last three lines of the model can be re-expanded using fully conjugate sampling schemes that do not require optimization or MCMC (only matrix operations).</p>
<p><strong>Here are the details:</strong> The collapsed model is given by <span class="math display">\[
\begin{align}
Y_j &amp; \sim \text{Multinomial}\left(\pi_j, n_j\right)  \\
\pi_j &amp; = \phi^{-1}(\eta_j) \\
\eta_j &amp;\sim T_{(D-1)\times N}(\upsilon, \Theta X, \Xi, I_N + X^T \Gamma X)
\end{align}
\]</span> where <span class="math inline">\(A=(I_N + X^T \Gamma, X)^{-1}\)</span> and <span class="math inline">\(T_{(D-1)\times N}\)</span> refers to the Matrix T-distribution the <span class="math inline">\((D-1)\times N\)</span> matrix <span class="math inline">\(\eta\)</span> with log density given by <span class="math display">\[\log T_{(D-1)\times N}(\eta | \upsilon, \Theta X, \Xi, A) \propto -\frac{\upsilon+N-D-2}{2}\log | I_{D-1}+\Xi^{-1}(\eta-\Theta X)A(\eta-\Theta X)^T |.\]</span> Rather than using MCMC to sample <span class="math inline">\(\eta\)</span> stray uses MAP estimation (using a custom C++ Eigen based implementation of the ADAM optimizer and closed form solutions for gradient and hessian of the collapsed model)<a href="#fn4" class="footnoteRef" id="fnref4"><sup>4</sup></a>. Additionally, <em>stray</em> allows quantification of uncertainty in MAP estimates using a Laplace approximation. We found that in practice this MAP based Laplace approximation produced comparable results to a full MCMC sampler but with tremendous improvements in compute time.</p>
<p>Once samples of <span class="math inline">\(\eta\)</span> are produced using the Laplace approximation closed form solutions for the conditional density of <span class="math inline">\(\Lambda\)</span> and <span class="math inline">\(\Sigma\)</span> given <span class="math inline">\(\eta\)</span> are used to “uncollapse” the collapsed model and produce posterior samples from the target model. This uncollapsing is fast and given by the following matrix equations:</p>
<p><span class="math display">\[
\begin{align}
\upsilon_N &amp;= \upsilon+N \\
\Gamma_N &amp;= (XX^T+\Gamma^{-1})^{-1} \\
\Theta_N &amp;= (\eta X^T+\Theta\Gamma^{-1})\Gamma_N \\
\Xi_N &amp;= \Xi + (\eta - \Theta_N X)(\eta - \Theta_N X)^T + (\Theta_N - \Theta)\Gamma(\Theta_N- \Theta)^T \\
p(\Sigma | \eta, X) &amp;= W^{-1}(\Xi_N, \upsilon_N)\\
p(\Lambda | \Sigma, \eta, X) &amp;= MN_{(D-1)\times Q}(\Lambda_N, \Sigma, \Gamma_N).
\end{align}
\]</span> If Laplace approximation is too slow, unstable (see below) or simply not needed, the default behavior of <em>pibble</em> is to preform the above matrix calculations and produce a single point estimate of <span class="math inline">\(\Sigma\)</span> and <span class="math inline">\(\Lambda\)</span> based on the posterior means of <span class="math inline">\(p(\Sigma | \eta, X)\)</span> and <span class="math inline">\((\Lambda | \Sigma, \eta, X)\)</span>.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-aitchison1986">
<p>Aitchison, J. 1986. <em>The Statistical Analysis of Compositional Data</em>. Book. Monographs on Statistics and Applied Probability. London ; New York: Chapman; Hall.</p>
</div>
<div id="ref-gevers2014">
<p>Gevers, Dirk, Subra Kugathasan, Lee A Denson, Yoshiki Vázquez-Baeza, Will Van Treuren, Boyu Ren, Emma Schwager, et al. 2014. “The Treatment-Naive Microbiome in New-Onset Crohn’s Disease.” <em>Cell Host &amp; Microbe</em> 15 (3). Elsevier: 382–92.</p>
</div>
<div id="ref-silverman2018">
<p>Silverman, Justin D, Heather Durand, Rachael J Bloom, Sayan Mukherjee, and Lawrence A David. 2018. “Dynamic Linear Models Guide Design and Analysis of Microbiota Studies Within Artificial Human Guts.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. doi:<a href="https://doi.org/10.1101/306597">10.1101/306597</a>.</p>
</div>
<div id="ref-silverman2019">
<p>Silverman, Justin D., Kimberly Roche, Zachary C. Holmes, Lawrence A. David, and Sayan Mukherjee. 2019. “Bayesian Multinomial Logistic Normal Models through Marginally Latent Matrix-T Processes.” <em>arXiv E-Prints</em>, March, arXiv:1903.11695.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>These are very large objects with many posterior samples, so it can take a little time to compute. Faster implementations of summary may be included as a future update if need arises<a href="#fnref1">↩</a></p></li>
<li id="fn2"><p>That said, due to the need to transform back and forth from the default coordinate system, it is fastest to call refit on <code>pibblefit</code> objects in the default coordinate system bypassing these transforms.<a href="#fnref2">↩</a></p></li>
<li id="fn3"><p>This can also be used to plot samples of the prior predictive distribution if Y is null in the object as in our <code>priors</code> object<a href="#fnref3">↩</a></p></li>
<li id="fn4"><p>Which we found preformed substantially better than L-BFGS, which we also tried.<a href="#fnref4">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#an-introduction-to-stray">An introduction to <em>stray</em></a><ul class="nav nav-pills nav-stacked">
<li><a href="#the-pibble-model">The <em>pibble</em> model</a></li>
      </ul>
</li>
      <li><a href="#example-analysis-of-microbiome-data">Example analysis of microbiome data</a></li>
      <li>
<a href="#more-technical-details">More Technical Details</a><ul class="nav nav-pills nav-stacked">
<li><a href="#a-few-notes-on-model-inference-and-parameter-collapsing">A few notes on model inference and parameter collapsing</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Justin Silverman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
